third/nmh/man/comp.man
.\"
.\" %nmhwarning%
.\" $Id: comp.man,v 1.1.1.1 1999-02-07 18:14:19 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH COMP %manext1% MH.6.8 [%nmhversion%]
.SH NAME
comp \- compose a message
.SH SYNOPSIS
.in +.5i
.ti -.5i
comp 
\%[+folder] \%[msg]
.br
.br
\%[\-form\ formfile]
\%[\-use] \%[\-nouse]
\%[\-file\ file]
.br
\%[\-draftfolder\ +folder]
\%[\-draftmessage\ msg]
.br
\%[\-nodraftfolder]
\%[\-editor\ editor]
\%[\-noedit]
.br
\%[\-whatnowproc\ program] \%[\-nowhatnowproc]
.br
\%[\-version]
\%[\-help]
.in -.5i
.SH DESCRIPTION
\fIComp\fR is used to create a new message to be mailed.  It copies a
message form to the draft being composed and then invokes an editor on
the draft (unless `\-noedit' is given, in which case the initial edit
is suppressed).

The default message form contains the following elements:

.nf
.in +.5i
.ne 10
.eo
.so %etcdir%/components
.ec
.in -.5i
.fi

If a file named \*(lqcomponents\*(rq exists in the user's nmh directory,
it will be used instead of this form.  You may specify an alternate
forms file with the switch `\-form\ formfile'.

You may also start \fIcomp\fR using the contents of an existing message
as the form.  If you supply either a `+folder' or `msg' argument, that
message will be used as the message form.  You may not supply both a
`\-form\ formfile' and a `+folder' or \&`msg' argument.  The line of
dashes or a blank line must be left between the header and the body of
the message for the message to be identified properly when it is sent
(see \fIsend\fR(1)).

The switch `\-use' directs \fIcomp\fR to continue editing an already
started message.  That is, if a \fIcomp\fR (or \fIdist\fR, \fIrepl\fR,
or \fIforw\fR\0) is terminated without sending the draft, the draft can
be edited again via \*(lqcomp\ \-use\*(rq.

The `\-file\ file' switch says to use the named file as the message draft.

If the draft already exists, \fIcomp\fR will ask you as to the disposition
of the draft.  A reply of \fBquit\fR will abort \fIcomp\fR, leaving
the draft intact; \fBreplace\fR will replace the existing draft with
the appropriate form; \fBlist\fR will display the draft; \fBuse\fR will
use the draft for further composition; and \fBrefile\ +folder\fR will
file the draft in the given folder, and give you a new draft with the
appropriate form.  (The `+folder' argument to \fBrefile\fR is required.)

The `\-draftfolder\ +folder' and `\-draftmessage\ msg' switches invoke
the \fInmh\fR draft folder facility.  This is an advanced (and highly
useful) feature.  Consult the \fImh-draft\fR(5) man page for more
information.

The `\-editor\ editor' switch indicates the editor to use for the
initial edit.  Upon exiting from the editor, \fIcomp\fR will invoke
the \fIwhatnow\fR program.  See \fIwhatnow\fR\0(1) for a discussion of
available options.  The invocation of this program can be inhibited
by using the `\-nowhatnowproc' switch.  (In truth of fact, it is
the \fIwhatnow\fR program which starts the initial edit.  Hence,
`\-nowhatnowproc' will prevent any edit from occurring.)
.Fi
^%etcdir%/components~^The standard message skeleton
^or <mh\-dir>/components~^Rather than the standard skeleton
^$HOME/\&.mh\(ruprofile~^The user profile
^<mh\-dir>/draft~^The draft file
.Pr
^Path:~^To determine the user's nmh directory
.Ps
^Draft\-Folder:~^To find the default draft\-folder
.Ps
^Editor:~^To override the default editor
.Ps
^Msg\-Protect:~^To set mode when creating a new message (draft)
.Ps
^fileproc:~^Program to refile the message
.Ps
^whatnowproc:~^Program to ask the \*(lqWhat now?\*(rq questions
.Sa
dist(1), forw(1), repl(1), send(1), whatnow(1), mh-profile(5)
.De
`+folder' defaults to the current folder
.Ds
`msg' defaults to the current message
.Ds
`\-nodraftfolder'
.Ds
`\-nouse'
.Co
None
.Bu
If \fIwhatnowproc\fR is \fIwhatnow\fR, then \fIcomp\fR uses a built\-in
\fIwhatnow\fR, it does not actually run the \fIwhatnow\fR program.
Hence, if you define your own \fIwhatnowproc\fR, don't call it
\fIwhatnow\fR since \fIcomp\fR won't run it.
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/comp.c

/*
 * comp.c -- compose a message
 *
 * $Id: comp.c,v 1.1.1.1 1999-02-07 18:14:13 danw Exp $
 */

#include <h/mh.h>
#include <fcntl.h>

static struct swit switches[] = {
#define	DFOLDSW                0
    { "draftfolder +folder", 0 },
#define	DMSGSW                 1
    { "draftmessage msg", 0 },
#define	NDFLDSW                2
    { "nodraftfolder", 0 },
#define	EDITRSW                3
    { "editor editor", 0 },
#define	NEDITSW                4
    { "noedit", 0 },
#define	FILESW                 5
    { "file file", 0 },
#define	FORMSW                 6
    { "form formfile", 0 },
#define	USESW                  7
    { "use", 0 },
#define	NUSESW                 8
    { "nouse", 0 },
#define	WHATSW                 9
    { "whatnowproc program", 0 },
#define	NWHATSW               10
    { "nowhatnowproc", 0 },
#define VERSIONSW             11
    { "version", 0 },
#define	HELPSW                12
    { "help", 4 },
    { NULL, 0 }
};

static struct swit aqrunl[] = {
#define	NOSW          0
    { "quit", 0 },
#define	YESW          1
    { "replace", 0 },
#define	USELSW        2
    { "use", 0 },
#define	LISTDSW       3
    { "list", 0 },
#define	REFILSW       4
    { "refile +folder", 0 },
#define NEWSW         5
    { "new", 0 },
    { NULL, 0 }
};

static struct swit aqrul[] = {
    { "quit", 0 },
    { "replace", 0 },
    { "use", 0 },
    { "list", 0 },
    { "refile", 0 },
    { NULL, 0 }
};


int
main (int argc, char **argv)
{
    int use = NOUSE, nedit = 0, nwhat = 0;
    int i, in, isdf = 0, out;
    char *cp, *cwd, *maildir, *dfolder = NULL;
    char *ed = NULL, *file = NULL, *form = NULL;
    char *folder = NULL, *msg = NULL, buf[BUFSIZ];
    char drft[BUFSIZ], **argp, **arguments;
    struct msgs *mp = NULL;
    struct stat st;

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* read user profile/context */
    context_read();

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    adios (NULL, "-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [+folder] [msg] [switches]",
			invo_name);
		    print_help (buf, switches, 1);
		    done (1);
		case VERSIONSW:
		    print_version(invo_name);
		    done (1);

		case EDITRSW: 
		    if (!(ed = *argp++) || *ed == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    nedit = 0;
		    continue;
		case NEDITSW: 
		    nedit++;
		    continue;

		case WHATSW: 
		    if (!(whatnowproc = *argp++) || *whatnowproc == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    nwhat = 0;
		    continue;
		case NWHATSW: 
		    nwhat++;
		    continue;

		case FORMSW: 
		    if (!(form = *argp++) || *form == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;

		case USESW: 
		    use++;
		    continue;
		case NUSESW: 
		    use = NOUSE;
		    continue;

		case FILESW: 	/* compatibility */
		    if (file)
			adios (NULL, "only one file at a time!");
		    if (!(file = *argp++) || *file == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    isdf = NOTOK;
		    continue;

		case DFOLDSW: 
		    if (dfolder)
			adios (NULL, "only one draft folder at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    dfolder = path (*cp == '+' || *cp == '@' ? cp + 1 : cp,
			    *cp != '@' ? TFOLDER : TSUBCWF);
		    continue;
		case DMSGSW: 
		    if (file)
			adios (NULL, "only one draft message at a time!");
		    if (!(file = *argp++) || *file == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;
		case NDFLDSW: 
		    dfolder = NULL;
		    isdf = NOTOK;
		    continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
	    else
		folder = path (cp + 1, *cp == '+' ? TFOLDER : TSUBCWF);
	} else {
	    if (msg)
		adios (NULL, "only one message at a time!");
	    else
		msg = cp;
	}
    }

    cwd = getcpy (pwd ());

    if (!context_find ("path"))
	free (path ("./", TFOLDER));

    /*
     * Check if we are using a draft folder
     * and have specified a message in it.
     */
    if ((dfolder || context_find ("Draft-Folder")) && !folder && msg && !file) {
	file = msg;
	msg = NULL;
    }
    if (form && (folder || msg))
	    adios (NULL, "can't mix forms and folders/msgs");

    if (folder || msg) {
	/*
	 * Use a message as the "form" for the new message.
	 */
	if (!msg)
	    msg = "cur";
	if (!folder)
	    folder = getfolder (1);
	maildir = m_maildir (folder);

	if (chdir (maildir) == NOTOK)
	    adios (maildir, "unable to change directory to");

	/* read folder and create message structure */
	if (!(mp = folder_read (folder)))
	    adios (NULL, "unable to read folder %s", folder);

	/* check for empty folder */
	if (mp->nummsg == 0)
	    adios (NULL, "no messages in %s", folder);

	/* parse the message range/sequence/name and set SELECTED */
	if (!m_convert (mp, msg))
	    done (1);
	seq_setprev (mp);	/* set the previous-sequence */

	if (mp->numsel > 1)
	    adios (NULL, "only one message at a time!");

	if ((in = open (form = getcpy (m_name (mp->lowsel)), O_RDONLY)) == NOTOK)
	    adios (form, "unable to open message");
    } else {
	/*
	 * Open a component or forms file
	 */
	if (form) {
	    if ((in = open (etcpath (form), O_RDONLY)) == NOTOK)
		adios (form, "unable to open form file");
	} else {
	    if ((in = open (etcpath (components), O_RDONLY)) == NOTOK)
		adios (components, "unable to open default components file");
	    form = components;
	}
    }

try_it_again:
    strncpy (drft, m_draft (dfolder, file, use, &isdf), sizeof(drft));

    /*
     * Check if we have an existing draft
     */
    if ((out = open (drft, O_RDONLY)) != NOTOK) {
	i = fdcompare (in, out);
	close (out);

	/*
	 * If we have given -use flag, or if the
	 * draft is just the same as the components
	 * file, then no need to ask any questions.
	 */
	if (use || i)
	    goto edit_it;

	if (stat (drft, &st) == NOTOK)
	    adios (drft, "unable to stat");
	printf ("Draft \"%s\" exists (%ld bytes).", drft, (long) st.st_size);
	for (i = LISTDSW; i != YESW;) {
	    if (!(argp = getans ("\nDisposition? ", isdf ? aqrunl : aqrul)))
		done (1);
	    switch (i = smatch (*argp, isdf ? aqrunl : aqrul)) {
		case NOSW: 
		    done (0);
		case NEWSW: 
		    file = NULL;
		    use = NOUSE;
		    goto try_it_again;
		case YESW: 
		    break;
		case USELSW:
		    use++;
		    goto edit_it;
		case LISTDSW: 
		    showfile (++argp, drft);
		    break;
		case REFILSW: 
		    if (refile (++argp, drft) == 0)
			i = YESW;
		    break;
		default: 
		    advise (NULL, "say what?");
		    break;
	    }
	}
    } else {
	if (use)
	    adios (drft, "unable to open");
    }

    if ((out = creat (drft, m_gmprot ())) == NOTOK)
	adios (drft, "unable to create");
    cpydata (in, out, form, drft);
    close (in);
    close (out);

edit_it:
    context_save ();	/* save the context file */

    if (nwhat)
	done (0);
    what_now (ed, nedit, use, drft, NULL, 0, NULLMP, NULL, 0, cwd);
    done (1);
}
